cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java

```diff
@@ -62,81 +64,139 @@
+   * RaftNodeAsClient list
    */
-  private final RaftNodeAsClient client = new RaftNodeAsClient();
-
-  private boolean clientInited = false;
+  private final LinkedList<RaftNodeAsClient> clientList = new LinkedList<>();
 
   /**
    * Number of clients in use
    */
-  private AtomicInteger validClientNum = new AtomicInteger(0);
+  private AtomicInteger clientNumInUse = new AtomicInteger(0);
 
   /**
    * Number of requests for clients in queue
    */
   private int queueClientNum = 0;
 
   /**
-   * Lock to update validClientNum
+   * Lock to update clientNumInUse
    */
-  private ReentrantLock numLock = new ReentrantLock();
+  private ReentrantLock resourceLock = new ReentrantLock();
+
+  /**
+   * Mark whether system is shutting down
+   */
+  private AtomicBoolean isShuttingDown = new AtomicBoolean(false);
 
   private RaftNodeAsClientManager(){
 
   }
 
+  public void init() {
+    isShuttingDown.set(false);
+  }
+
   /**
-   * Try to get client, return null if num of queue client exceeds threshold.
+   * Try to get clientList, return null if num of queue clientList exceeds threshold.
    */
-  public RaftNodeAsClient getRaftNodeAsClient() {
+  public RaftNodeAsClient getRaftNodeAsClient() throws RaftConnectionException {
     try {
-      if (!clientInited) {
-        client.init();
-      }
-
-      numLock.lock();
-      if (validClientNum.get() < MAX_VALID_CLIENT_NUM) {
-        validClientNum.incrementAndGet();
-        return client;
-      }
+      resourceLock.lock();
       if (queueClientNum >= MAX_QUEUE_CLIENT_NUM) {
-        return null;
+        throw new RaftConnectionException(String
+            .format("Raft inner rpc clients have reached the max numbers %s",
+                CLUSTER_CONFIG.getMaxNumOfInnerRpcClient() + CLUSTER_CONFIG
+                    .getMaxQueueNumOfInnerRpcClient()));
+      }
+      checkShuttingDown();
+      if (clientNumInUse.get() < MAX_VALID_CLIENT_NUM) {
+        clientNumInUse.incrementAndGet();
+        return getClient();
       }
       queueClientNum++;
     }finally {
-      numLock.unlock();
+      resourceLock.unlock();
     }
     return tryToGetClient();
   }
 
+  private void checkShuttingDown() throws RaftConnectionException {
+    if (isShuttingDown.get()) {
+      throw new RaftConnectionException(
+          "Reject to provide RaftNodeAsClient client because cluster system is shutting down");
+    }
+  }
+
   /**
-   * Check whether it can get the client
+   * Check whether it can get the clientList
    */
-  private RaftNodeAsClient tryToGetClient() {
+  private RaftNodeAsClient tryToGetClient() throws RaftConnectionException {
     for(;;){
-      if(validClientNum.get() < MAX_VALID_CLIENT_NUM){
+      if(clientNumInUse.get() < MAX_VALID_CLIENT_NUM){
+        resourceLock.lock();
         try{
-          numLock.lock();
-          if(validClientNum.get() < MAX_VALID_CLIENT_NUM){
-            validClientNum.incrementAndGet();
+          checkShuttingDown();
```

Comment:
If queue is full, try sleep(1000) is better? `for(;;)` wastes cpu time

Reply:
fixed